### PR TITLE
Move the extended lifetime resolution into typeck context

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
Related to #15023

This PR is based on the [idea](https://github.com/rust-lang/rust/issues/15023#issuecomment-1070931433) of #15023 by @nikomatsakis.

This PR specifically proposes to
- Delay the resolution of scopes of rvalues to a later stage, so that enough type information is available to refine those scopes based on relationships of lifetimes.
- Highlight relevant parts that would help future reviews on the next installments of works to fully implement a solution to RFC 66.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -367,6 +368,11 @@ pub struct GeneratorInteriorTypeCause<'tcx> {
     pub expr: Option<hir::HirId>,
 }
 
+#[derive(TyEncodable, TyDecodable, Clone, Debug, Eq, PartialEq, HashStable)]
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/ty/context.rs:371]: This deserves a comment for sure =)

File: compiler/rustc_middle/src/ty/context.rs [hunk 2]
```diff
@@ -367,6 +368,11 @@ pub struct GeneratorInteriorTypeCause<'tcx> {
...
+pub struct RvalueScopes {
+    pub inner: FxHashMap<hir::ItemLocalId, Option<Scope>>,
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/ty/context.rs:373]:
```suggestion
    map: FxHashMap<hir::ItemLocalId, Option<Scope>>,
```

Nit: I'd move the things that manipulate the map to methods on `RvalueScopes` (and move this definition over to where its `impl` is)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Left some nits, but the bigger question is: can we move the "rvalue candidate" stuff into the same place that the scopes are computed? (typeck)

I think we will want that eventually, because the candidates will be influenced by the results of type check.

File: compiler/rustc_typeck/src/check/rvalue_scopes.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+pub fn resolve_rvalue_scopes<'a, 'tcx>(fcx: &'a FnCtxt<'a, 'tcx>, def_id: DefId) -> RvalueScopes {
+    let tcx = &fcx.tcx;
+    let hir_map = tcx.hir();
+    let mut rvalue_scopes = RvalueScopes::new();
+    debug!("start resolving rvalue scopes, def_id={def_id:?}");
+    let scope_tree = tcx.region_scope_tree(def_id);
+    debug!("rvalue_scope: rvalue_candidates={:?}", scope_tree.rvalue_candidates);
+    for (&hir_id, candidate) in &scope_tree.rvalue_candidates {
```
[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply compiler/rustc_typeck/src/check/rvalue_scopes.rs:73]: I have tried to move it into `typeck`. There is an issue that I am not sure about. This involves walking the HIR and the code would be probably mirroring exactly what the `rustc_passes::region::region_scope_tree` is doing. I am afraid that there will be some code duplication. If you think this is necessary, I will give it another try.

[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply compiler/rustc_typeck/src/check/rvalue_scopes.rs:73]: Specifically, the visitor is computing a few things at the same time. The most important information that `rvalue_candidates` relies on is the parent scope information. Apparently it is also largely connected to computing other scope information like `var_scopes` and sorts.

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply compiler/rustc_typeck/src/check/rvalue_scopes.rs:73]: Hmm. I'll have to investigate.

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply compiler/rustc_typeck/src/check/rvalue_scopes.rs:73]: Any chance you can schedule a session at https://calendly.com/nikomatsakis/office-hours and we can investigate together?

[the remark this answers is not in the document]
Comment (dingxiangfei2009) [reply compiler/rustc_typeck/src/check/rvalue_scopes.rs:73]: Yes, very glad to. I have booked a session with calendly.

File: compiler/rustc_middle/src/ty/context.rs (hunk 2 above)
Comment (dingxiangfei2009) [reply compiler/rustc_middle/src/ty/context.rs]: Yes, I think making it `pub` was actually not necessary.

File: compiler/rustc_middle/src/middle/region.rs [hunk 4]
```diff
@@ -223,15 +223,10 @@ pub struct ScopeTree {
...
-    /// expression ID to the cleanup scope id. For rvalues not present in
-    /// this table, the appropriate cleanup scope is the innermost
-    /// enclosing statement, conditional expression, or repeating
-    /// block (see `terminating_scopes`).
-    /// In constants, None is used to indicate that certain expressions
-    /// escape into 'static and should have no local cleanup scope.
-    rvalue_scopes: FxHashMap<hir::ItemLocalId, Option<Scope>>,
+    /// Identifies expressions which, if captured into a temporary, ought to
+    /// have a temporary whose lifetime extends to the end of the enclosing *block*,
+    /// and not the enclosing *statement*. (See `RvalueCandidate` for more information.)
```
Comment (nikomatsakis) [inline compiler/rustc_middle/src/middle/region.rs:228]:
```suggestion
    /// Identifies expressions which, if captured into a temporary, ought to
    /// have a temporary whose lifetime extends to the end of the enclosing *block*,
    /// and not the enclosing *statement*. Expressions that are not present in this
    /// table are not rvalue candidates. The set of rvalue candidates is computed
    /// during type check based on a traversal of the AST.
```

**Review by nikomatsakis — APPROVED** [review]
Yes, this looks great!

**Comment by nikomatsakis** [comment]
@bors delegate+

@dingxiangfei2009 This is r=me but I can't r+ because it needs a rebase. I'm delegating to you so that you can rebase and then do `bors r+`

**Comment by Urgau** [comment]
> @bors r+

Heu, it should have been `@bors r=nikomatsakis`; You can't approve your own PR.

I suggest doing `@bors r-` and then `@bors r=nikomatsakis` (or maybe just the last, don't know bors enough).

**Comment by dingxiangfei2009** [comment]
@bors r-
@bors r=nikomatsakis

Let me try this again. :sweat_smile:

**Comment by Mark-Simulacrum** [comment]
@nikomatsakis @dingxiangfei2009 It looks like this PR was a regression across a fairly large number of incremental benchmarks -- I don't see a previous perf run done on this PR or discussion of performance as part of this change. Was this regression expected at an architectural level?

It looks like non-incremental benchmarks actually improved, so perhaps we're not caching something we were before?

**Comment by dingxiangfei2009** [comment]
I have a question now. Is there a way to cache things by keys of type `DefId` without queries? Or keeping it as a query is the way to go?

**Comment by pnkfelix** [comment]
In addition to the instruction count regression, it seems like it also caused a regression to the compiler's memory usage (max-rss): 

https://perf.rust-lang.org/compare.html?start=acfd327fd4e3a302ebb0a077f422a527a7935333&end=653463731a7f01f519cf85f444869def27f00395&stat=max-rss

Will restoring the query also address the max-rss regression?a

**Pull request merged**